cmd/clusterctl/pkg/client/cluster/inventory.go

```diff
@@ -294,3 +308,115 @@ func (p *inventoryClient) GetDefaultProviderNamespace(provider string) (string,
 	// There is no provider or more than one namespace for this provider; in both cases, a default provider namespace cannot be decided.
 	return "", nil
 }
+
+func (p *inventoryClient) GetManagementGroups() ([]ManagementGroup, error) {
+	providers, err := p.List()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if there are core providers with overlapping watching namespaces, if yes, it is not possible to define the management groups, so return error
+	if err := checkOverlappingCoreProviders(providers); err != nil {
+		return nil, err
+	}
+
+	// check if a Bootstrap/ControlPlane/Infrastructure providers can't be combined with any core provider or
+	// if any Bootstrap/ControlPlane/Infrastructure provider is watching for objects controlled by more than one core provider.
+	// In both cases it is not possible to define the management groups, so return error
+	if err := checkOverlappingProviders(providers); err != nil {
+		return nil, err
+	}
+
+	managementGroups := []ManagementGroup{}
+	for _, coreProvider := range providers {
+		if coreProvider.Type != string(clusterctlv1.CoreProviderType) {
+			continue
+		}
+
+		group := ManagementGroup{CoreProvider: coreProvider}
+		for _, provider := range providers {
+			if coreProvider.WatchedNamespace == "" || coreProvider.WatchedNamespace == provider.WatchedNamespace || provider.WatchedNamespace == "" {
+				group.Providers = append(group.Providers, provider)
```

Comment:
Just to double check: we're not removing the providers selected in a management group because there should never be any overlap given the checks above? I'd add a note here

Reply:
I'm not 100 sure I understand the question. 

In a Management group there are _different_ providers that cooperate in reconciling objects.
So technically those _different_ providers are expected to overlap.

What is checked above is if more instances of the  _same_ provider are overlapping, which is not allowed